cfgmgr/buffermgrdyn.cpp

```diff
@@ -0,0 +1,1674 @@
+{
+    // Crete record in BUFFER_PROFILE table
+    // key format is pg_lossless_<speed>_<cable>_profile
+    vector<FieldValueTuple> fvVector;
+    auto &profile = m_bufferProfileLookup[profile_name];
+
+    if (profile.static_configured)
+    {
+        // Check whether the profile is statically configured.
+        // This means:
+        // 1. It's a statically configured profile, headroom override
+        // 2. It's dynamically calculated headroom with static threshold (alpha)
+        // In this case we won't remove the entry from the local cache even if it's dynamically calculated
+        // because the speed, cable length and cable model are fixed the headroom info will always be valid once calculated.
+        SWSS_LOG_NOTICE("Unable to release profile %s because it's a static configured profile", profile_name.c_str());
+        return;
+    }
+
+    // Check whether it's referenced anymore by other PGs.
+    if (!profile.port_pgs.empty())
+    {
+        for (auto &pg : profile.port_pgs)
+        {
+            SWSS_LOG_INFO("Unable to release profile %s because it's still referenced by %s (and others)",
+                          profile_name.c_str(), pg.c_str());
+            return;
+        }
+    }
+
+    profile.port_pgs.clear();
+
+    m_applBufferProfileTable.del(profile_name);
+
+    m_stateBufferProfileTable.del(profile_name);
+
+    m_bufferProfileLookup.erase(profile_name);
+
+    SWSS_LOG_NOTICE("BUFFER_PROFILE %s has been released successfully", profile_name.c_str());
+}
+
+bool BufferMgrDynamic::isHeadroomResourceValid(const string &port, const buffer_profile_t &profile, const string &new_pg = "")
+{
+    //port: used to fetch the maximum headroom size
+    //profile: the profile referenced by the new_pg (if provided) or all PGs
+    //new_pg: which pg is newly added?
+
+    if (!profile.lossless)
+    {
+        SWSS_LOG_INFO("No need to check headroom for lossy PG port %s profile %s size %s pg %s",
+                  port.c_str(), profile.name.c_str(), profile.size.c_str(), new_pg.c_str());
+        return true;
+    }
+
+    bool result = true;
+
+    vector<string> keys = {port};
+    vector<string> argv = {};
+
+    argv.emplace_back(profile.name);
+    argv.emplace_back(profile.size);
+
+    if (!new_pg.empty())
+    {
+        argv.emplace_back(new_pg);
+    }
+
+    SWSS_LOG_INFO("Checking headroom for port %s with profile %s size %s pg %s",
+                  port.c_str(), profile.name.c_str(), profile.size.c_str(), new_pg.c_str());
+
+    try
+    {
+        auto ret = runRedisScript(*m_applDb, m_checkHeadroomSha, keys, argv);
+
+        // The format of the result:
+        // a list of strings containing key, value pairs with colon as separator
+        // each is the size of a buffer pool
+
+        for ( auto i : ret)
+        {
+            auto pairs = tokenize(i, ':');
+            if ("result" == pairs[0])
+            {
+                if ("true" != pairs[1])
```

Comment:
which method checks for the valid and available headroom and update the appDb??

Reply:
The lua plugin `buffer_check_headroom_<vendor>` will check it.
After the check passed, the PG will be updated by `updateBufferPgToDb`.
I believe the big picture of the logic can be found in [refreshPriorityGroupsForPort](https://github.com/Azure/sonic-swss/pull/1338/files/80b5b1f1b4dbf7b16415882911cf7253f86c35f5#diff-908ebd83566b58a4ad7eb1133cac45f1fa4f328ada69180a47e496b25a34cee7R670)